Automattic/newspack-plugin
# React Router + Wizard Higher Order Components

### All Submissions:






### Changes proposed in this Pull Request:



This PR aims to achieve the goals of https://github.com/Automattic/newspack-plugin/pull/104 using a few different constructs. The functionality should be almost identical to 104, with a few minor differences noted below.

1) Routing 
All of the home-grown Wizard navigation has been replaced by routing, using [React Router](https://reacttraining.com/react-router/). I created a separate PR (not meant to  be merged) with only the routing piece, to make it a bit easier to assess how it works: https://github.com/Automattic/newspack-plugin/pull/107. Originally we had thought that Routing was a future nice-to-have, but a couple of factors changed my mind:

- The routing work in [104](https://github.com/Automattic/newspack-plugin/pull/104) is pretty deep and complicated (all the `identifier` stuff), which made this feel like a now-or-never kind of moment for routing.
- As part of the recent `Handoff` button work, the last visited Newspack URL is persisted so that the user can return from other plugins to exactly where they left off. Routing allows this to work with much greater precision, e.g. being able to return to the edit page of a specific Subscription.
- There's been discussion that the entire plugin might one day be one single page app as opposed to a series of different ones. This transition will be much easier if we are already using a standard routing strategy. 
- Routing improves the overall UI in a few ways, e.g. browser back button and history.
- React Router appears to be a widely used, well supported library. @iuravic has used it previously and spoke highly of it. 

2) Higher Order Components
Instead of the `Wizard` and `WizardScreen` components, I created two higher-order components: `withWizard` and `withWizardScreen`. These basically do  what the components did, wrapping components in another one which inject props, alter the render, and provide additional functionality. 

`withWizard` provides plugin requirement checking, error handling, fires `onWizardReady` in the wrapped component, and adds an optional button at the bottom.

`withWizardScreen` renders `<FormattedHeader/>`, wraps the screen in a `<Card/>` and renders an optional `<Button/>` at the bottom. 

## Differences From 104

1) In PR 104, `WizardScreen` renders primary and secondary buttons. Since the secondary button is generally going to be used to return to the relevant checklist screen, I opted to move this to `Wizard`. So each HOC now has one built-in button.

2) With routing introduced, `<Button/>` elements now frequently require an `href` rather than an `onClick` event. Rather than taking separate attributes for events and URLs, I added a utility `buttonProps` which will create a set of props for a `<Button/>` element. If the value is a function the prop will be an `onClick` event. If the value is a string it will be an `href`.

3) The `Wizard`/`WizardScreen` demo on Components Demo has been removed entirely. To make a demo of the new structure we would need to create a few new blank components for that purpose only, which felt wasteful. 

4) For reference, the following files have been copied from 104 verbatim:
`assets/components/src/with-wizard-screen/style.scss`
`assets/wizards/subscriptions/style.scss`
`includes/class-plugin-manager.php`
`includes/wizards/class-subscriptions-wizard.php`

## Two Failed Experiments
In the spirit of completeness, here are two other approaches I tried which didn't work out:

1) Initially I attempted to take PR 104 as-is and implement routing entirely within the `Wizard` and `WizardScreen` components, thus leaving the implementations in the wizards untouched. This work is here: https://github.com/Automattic/newspack-plugin/pull/106. I got pretty far with it, but it ultimately became utterly convoluted. There were lots of situations where I needed to change routes programmatically from components where this wasn't expected behavior and I ended up with something pretty messy... Eventually I abandoned the approach in favor of a "normal" implementation of React Router.

2) I experimented with keeping the `<HashRouter>` and `<Switch>` wrapping in `withWizard`. This would have been nice because each Wizard's render would only need to contain a series of `<Route/>`s, and the Plugin Management UI could be injected automatically. The approach worked, but broke the React `ref` I'm using to call `onWizardComplete` after plugin management is complete. I'm not positive why this was, but ultimately abandoned the approach. All that this means is that `<HashRouter>`, `<Switch>`, and `{ pluginRequirements }` must be included in every Wizard's render, which isn't a terribly great sacrifice.

### How to test the changes in this Pull Request:

1. `npm ci && npm run clean && npm run build:webpack -- --watch`
2. Test all functionality in the Subscription Onboarding and Subscriptions wizards. These should be effectively identical to how they work on `master`
3. Observe URL changes to verify routing. 

### Other information:
